datastore/tests/unit/filter_test.py

```diff
@@ -0,0 +1,165 @@
+                    }
+                },
+                {
+                    'propertyFilter': {
+                        'property': {
+                            'name': original_filter.filters[1].inner_filter.prop
+                        },
+                        'op': original_filter.filters[1].inner_filter.operator,
+                        'value': original_filter.filters[1].inner_filter.value.to_repr()
+                    }
+                },
+            ]
+        }
+
+        output_filter = CompositeFilter.from_repr(data)
+
+        assert output_filter == original_filter
+
+    def test_composite_filter_to_repr(self):
+        property_filters = self._create_property_filters(count=2)
+        composite_filter = CompositeFilter(
+            operator=CompositeFilterOperator.AND,
+            filters=[
+                Filter(property_filters[0]),
+                Filter(property_filters[1])
+            ])
+        query_filter = Filter(composite_filter)
+
+        r = query_filter.to_repr()
+
+        composite_filter_dict = r['compositeFilter']
+        assert composite_filter_dict['op'] == 'AND'
+        self._assert_is_correct_prop_dict_for_property_filter(
+            composite_filter_dict['filters'][0]['propertyFilter'], property_filters[0])
+        self._assert_is_correct_prop_dict_for_property_filter(
+            composite_filter_dict['filters'][1]['propertyFilter'], property_filters[1])
+
+    def test_filter_from_repr(self):
+        composite_filter = self._create_composite_filter()
+        original_filter = Filter(inner_filter=composite_filter)
+
+        data = {
+            'compositeFilter': original_filter.inner_filter.to_repr()
+        }
+
+        output_filter = Filter.from_repr(data)
+
+        assert output_filter == original_filter
+
+    def test_filter_from_repr_unexpected_filter_name(self):
+        unexpected_filter_name = 'unexpectedFilterName'
+        data = {
+            unexpected_filter_name: 'DoesNotMatter'
+        }
+
+        with pytest.raises(ValueError) as ex_info:
+            Filter.from_repr(data)
+
+        assert unexpected_filter_name in ex_info.value.args[0]
+
+    def test_filter_to_repr(self):
+        composite_filter = self._create_composite_filter()
+        test_filter = Filter(inner_filter=composite_filter)
+
+        r = test_filter.to_repr()
+
+        assert r['compositeFilter'] == test_filter.inner_filter.to_repr()
+
+    def test_repr_returns_to_repr_as_string(self):
+        query_filter = self._create_filter()
+
+        assert repr(query_filter) == str(query_filter.to_repr())
+
+    @staticmethod
+    def _create_property_filter() -> PropertyFilter:
+        return TestFilter._create_property_filters(count=1)[0]
+
+    @staticmethod
+    def _create_property_filters(count: int) -> List[PropertyFilter]:
+        if count > 2:
+            raise NotImplementedError('You may need to implement a way to generate any number of property filters :)')
```

Comment:
This might look a bit sloppy but it gets the benefits of a [Creation Method](http://xunitpatterns.com/Creation%20Method.html) that works very nicely with the tests above, without writing a method that is more general than I need (YAGNI). The day you actually need that, this message will let you know :)

Also, it just occurred to me that pylint didn't complain about this line being too long. I think it's either because it doesn't check test files or it's related to the fact that its code is raising an exception on my machine.

Reply:
Interesting, I hadn't seen this before! What's the benefit to using these magic methods and property creation methods rather than using the standard test fixture workflow?

For example:
```
@pytest.fixture()
def property_filters() -> List[PropertyFilter]:
    return [
            PropertyFilter(prop='prop1', operator=PropertyFilterOperator.LESS_THAN, value=Value('value1')),
            PropertyFilter(prop='prop2', operator=PropertyFilterOperator.GREATER_THAN, value=Value(1234)),
        ]

@pytest.fixture()
def compose_filter(property_filters) -> CompositeFilter:
            return CompositeFilter(
            operator=CompositeFilterOperator.AND,
            filters=[
                Filter(property_filters[0]),
                Filter(property_filters[1])
            ])
```

which can be used in tests with:

```
def test_filter_to_repr(composite_filter):
    test_filter = Filter(inner_filter=composite_filter)
    r = test_filter.to_repr()
    assert r['compositeFilter'] == test_filter.inner_filter.to_repr()
```

Reply:
None! I'm not used to pytest because we use unittest in the other projects.
But this syntax looks great, I'll have a look and update the tests.